### ISSUE TITLE
feat: extract LoRA for arbitrary models

### ISSUE THREAD

**Issue by jon-chuang** [huggingface/peft#312]
It's straightforward, using [low-rank approximation](https://en.wikipedia.org/wiki/Low-rank_approximation), [low-rank matrix factorization](https://www.sciencedirect.com/science/article/abs/pii/S0925231217315710#:~:text=The%20purpose%20of%20low%2Drank,the%20interactions%20between%20two%20entries.).

Given a model and its fine-tuning, and a target rank $k$, extract the "best" low-rank approximation to each difference in the model weights, and export as LoRA.

The parameter $k$ can be a constant or can be unique to each matrix, e.g. $k_i$ for $W_i \in \Theta$

To summarize, given $W' - W = \Delta W$, find $\hat{A},\hat{B}$ each with $k$ rows such that the norm $||\hat{A}^T\hat{B} - \Delta W||_2$ is minimized.

**Comment by mekaneeky** [comment]
This is definitely interesting. Wonder whether gradient approaches, evolutionary algorithms or plain old linear algebra norms (Spectral, etc) and factorization would be ideal for solving this.

**Comment by johnwick123f** [comment]
Hmm yeah I would like this feature too.

**Comment by thomasgauthier** [comment]
I implemented something in this direction using singular value decomposition (SVD). I call it [LoRD](https://github.com/thomasgauthier/LoRD) for *Lo*w-*R*ank *D*ecomposition
